### PR TITLE
PLG-835: Use ProductId and ProductModelId value objects in DQI

### PR DESCRIPTION
**Pull request by olivierpontier**
**Description (for Contributor and Core Developer)**

In order to identify the products with a new UUID property, we need to clean the way we compute and retrieve scores and evaluations with the ProductId value object. We want to use ProductId when it concerns Product and ProductModelId when it concerns ProcuctModel

- [x] unit tests
- [x] integration tests
- [x] replace by a factory service in charge of creating ProductId or ProductModelId

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Enrichment/CalculateProductCompleteness.php [hunk 1]
```diff
@@ -25,14 +25,18 @@ final class CalculateProductCompleteness implements \Akeneo\Pim\Automation\DataQ
...
     ) {
         $this->completenessCalculator = $completenessCalculator;
         $this->getProductIdentifierFromProductIdQuery = $getProductIdentifierFromProductIdQuery;
     }
 
-    public function calculate(ProductId $productId): CompletenessCalculationResult
+    public function calculate(ProductEntityIdInterface $productId): CompletenessCalculationResult
     {
+        if (!$productId instanceof ProductId) {
+            throw new \InvalidArgumentException(sprintf('Invalid product model id: %s', (string)$productId));
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Enrichment/CalculateProductCompleteness.php:37]:
```suggestion
            throw new \InvalidArgumentException(sprintf('Invalid product id: %s', (string)$productId));
```

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Symfony/Command/PopulateProductModelScoresAndKeyIndicatorsCommand.php [hunk 2]
```diff
@@ -60,7 +61,7 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 
         while ($productModelIds = $this->getNextProductModelIds($lastProductModelId)) {
             try {
-                $productModelIdsCollection = ProductIdCollection::fromInts($productModelIds);
+                $productModelIdsCollection = $this->idFactory->createCollection(array_map(fn ($productId) => (string) $productId, $productModelIds));
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Symfony/Command/PopulateProductModelScoresAndKeyIndicatorsCommand.php:64]:
```suggestion
                $productModelIdsCollection = $this->idFactory->createCollection(array_map(fn ($productModelId) => (string) $productModelId, $productModelIds));
```

**Review by nmarniesse — APPROVED** [review]
Good job!

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/ValueObject/ProductModelIdCollection.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+final class ProductModelIdCollection implements ProductEntityIdCollection
+{
+    /**
+     * @var array<ProductEntityIdInterface>
+     */
+    private array $productIds;
+
+    private function __construct(array $productIds)
+    {
+        Assert::allIsInstanceOf($productIds, ProductModelId::class);
```
Comment (LaurentPetard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/ValueObject/ProductModelIdCollection.php:23]: The constructor is private, and the class final, so it's useless to check the instances of `$productsIds` here.
It's enough to ensure that we have the good instance in the public methods

**Review by LaurentPetard — APPROVED** [review]

**Review by FSAkeneo — APPROVED** [review]

**Pull request merged**